### PR TITLE
Expose AsyncMap keys, values and entries

### PR DESCRIPTION
**Pull request by tsegismont**
See https://github.com/eclipse/vert.x/issues/2020

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you please review?

File: src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * @author Thomas Segismont
+ */
+public class CloseableIteratorCollectionStream<IN, OUT> implements ReadStream<OUT> {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java]: I / O it's Java :-)

File: src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  @Override
+  public synchronized CloseableIteratorCollectionStream<IN, OUT> endHandler(Handler<Void> handler) {
+    endHandler = handler;
+    return this;
+  }
+
+  private void close() {
+    closed = true;
+    context.executeBlocking(fut -> {
+      if (iterator != null) {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java]: iterator should be accessed under synchronization

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java [hunk 3]
```diff
@@ -57,4 +69,127 @@ protected void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>>
   protected ClusterManager getClusterManager() {
     return new InfinispanClusterManager();
   }
+
+  @Test
+  @Repeat(times = 10)
```
Comment (vietj) [inline src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: good idea

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java [hunk 4]
```diff
@@ -57,4 +69,127 @@ protected void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>>
...
+  public void testKeyStream() {
+    Map<JsonObject, Buffer> map = genJsonToBuffer(100);
+    loadData(map, (vertx, asyncMap) -> {
+      List<JsonObject> keys = new ArrayList<>();
+      ReadStream<JsonObject> stream = InfinispanAsyncMap.<JsonObject, Buffer>unwrap(asyncMap).keyStream();
+      long pause = 500;
+      Long start = System.nanoTime();
```
Comment (vietj) [inline src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: why `Long` ?

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java [hunk 5]
```diff
@@ -57,4 +69,127 @@ protected void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>>
...
+          });
+        }
+      });
+    });
+    await();
+  }
+
+  @Test
+  @Repeat(times = 10)
+  public void testEntryStream() {
```
Comment (vietj) [inline src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: three tests look very similar, perhaps you can make one generic test and use a variable to customize it ?

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java (hunk 3 above)
Comment (tsegismont) [reply src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: Actually a left-over from my local testing. Do we want to keep it?

File: src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java (hunk 1 above)
Comment (tsegismont) [reply src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java]: Done

File: src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java (hunk 2 above)
Comment (tsegismont) [reply src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java]: Done

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java (hunk 4 above)
Comment (tsegismont) [reply src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: My bad. Updated

File: src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java (hunk 5 above)
Comment (tsegismont) [reply src/test/java/io/vertx/ext/cluster/infinispan/test/InfinispanClusterWideMapTest.java]: Done

**Comment by tsegismont** [comment]
@vietj PR updated

**Pull request merged**